File: test/test-echo.rb
#!/usr/bin/env ruby
# coding: utf-8
require 'yaml'
require 'test/unit'
require_relative '../cwl/inspector'

unless defined? CWL_PATH
  CWL_PATH=File.join(File.dirname(__FILE__), '..', 'examples')
end

class TestEcho < Test::Unit::TestCase
  def setup
    @cwl = File.join(CWL_PATH, 'echo', 'echo.cwl')
    @runtime = {
      'outdir' => File.absolute_path('.'),
      'docdir' => '.',
      'tmpdir' => '/tmp',
    }
    @vardir = case RUBY_PLATFORM
              when /darwin|mac os/
                '/private/var'
              when /linux/
                '/var'
              else
                raise "Unsupported platform: #{RUBY_PLATFORM}"
              end
    @use_docker = system('which docker > /dev/null')
  end

  def test_version
    assert_equal('v1.0', walk(@cwl, '.cwlVersion'))
  end

  def test_id_based_access
    assert_equal('Input string', walk(@cwl, '.inputs.input.label'))
  end

  def test_index_based_access
    assert_equal('Input string', walk(@cwl, '.inputs.0.label'))
  end

  def test_commandline
    cmd = if @use_docker
            "docker run -i --read-only --rm --workdir=#{@vardir}/spool/cwl --env=HOME=#{@vardir}/spool/cwl --env=TMPDIR=/tmp --user=#{Process::UID.eid}:#{Process::GID.eid} -v #{Dir.pwd}:#{@vardir}/spool/cwl -v /tmp:/tmp docker/whalesay \"cowsay\"  > #{Dir.pwd}/output"
          else
            sh = case RUBY_PLATFORM
                 when /darwin|mac os/
                   '/bin/bash'
                 else
                   '/bin/sh'
                 end
            "env HOME='#{Dir.pwd}' TMP='#{@runtime['tmpdir']}' #{sh} -c 'cd ~ && \"cowsay\" ' > #{Dir.pwd}/output"
          end
    assert_equal(cmd, commandline(@cwl, @runtime, parse_inputs(@cwl, {}, @runtime)))
  end

  def test_instantiated_commandline
    cmd = if @use_docker
            "docker run -i --read-only --rm --workdir=#{@vardir}/spool/cwl --env=HOME=#{@vardir}/spool/cwl --env=TMPDIR=/tmp --user=#{Process::UID.eid}:#{Process::GID.eid} -v #{Dir.pwd}:#{@vardir}/spool/cwl -v /tmp:/tmp docker/whalesay \"cowsay\" \"Hello!\" > #{Dir.pwd}/output"
          else
            "env HOME='#{Dir.pwd}' TMP='#{@runtime['tmpdir']}' #{sh} -c 'cd ~ && \"cowsay\" \"Hello!\"' > #{Dir.pwd}/output"
          end
    assert_equal(cmd, commandline(@cwl, @runtime,
                                  parse_inputs(@cwl, { 'input' => 'Hello!' }, @runtime)))
  end

  def test_root_keys
    assert_equal(['baseCommand', 'class', 'cwlVersion', 'hints', 'id',
                  'inputs', 'outputs', 'stdout', 'successCodes'],
                 keys(@cwl, '.').sort)
  end

  def test_keys
    assert_equal(['input'], keys(@cwl, '.inputs'))
  end
end

File: test/test_echo.py
#!/usr/bin/env python
import unittest
from os.path import basename
from cwl_parser import load_document
from cwl_inspector import inspect

CWL = 'examples/echo/echo.cwl'


class TestEcho(unittest.TestCase):

    def setUp(self):
        self.cwl = load_document(CWL)

    def test_version(self):
        self.assertEqual('v1.0', inspect(self.cwl, '.cwlVersion'))

    def test_ID_based_access(self):
        self.assertEqual('Input string',
                         inspect(self.cwl, '.inputs.input.label'))

    def test_index_based_access(self):
        self.assertEqual('Input string',
                         inspect(self.cwl, '.inputs.0.label'))

    def test_commandline(self):
        pass

    def test_instantiated_commandline(self):
        pass

    def test_root_keys(self):
        self.assertEqual(['arguments', 'baseCommand', 'class_', 'cwlVersion',
                          'doc', 'hints', 'id', 'inputs', 'label', 'outputs',
                          'permanentFailCodes', 'requirements', 'stderr',
                          'stdin', 'stdout', 'successCodes',
                          'temporaryFailCodes'],
                         inspect(self.cwl, 'keys(.)'))

    def test_keys(self):
        self.assertEqual(['input'],
                         [basename(f) for f in inspect(self.cwl,
                                                       'keys(.inputs)')])
